File: kid-ios/MyWorld/Views/RolePickerView.swift
import SwiftUI

/// First-run chooser: is this the child's board or the parent's device?
/// Shown once after login when no role is stored. Both choices are reversible
/// from settings — picking wrong is a two-tap fix, so no confirmation step.
struct RolePickerView: View {
    @Environment(DeviceMode.self) private var mode
    @Environment(AuthManager.self) private var auth
    /// The tapped role while the destination view spins up — the first load
    /// pulls the whole board over the network, which can take several seconds
    /// with NOTHING on screen unless we say so here.
    @State private var opening: DeviceMode.Role?

    var body: some View {
        VStack(spacing: 28) {
            Spacer()
            Image("MyWorldLogo")
                .resizable()
                .scaledToFit()
                .frame(width: 84, height: 84)
                .clipShape(RoundedRectangle(cornerRadius: 19, style: .continuous))
                .shadow(color: .black.opacity(0.12), radius: 8, y: 3)
            Text("Who uses this device?")
                .font(.system(size: 30, weight: .bold, design: .rounded))
                .foregroundStyle(Color(hex: "#ad1457"))

            VStack(spacing: 16) {
                roleButton(
                    icon: "hand.tap.fill",
                    title: "\(childPossessive(auth.user?.slug, fallback: "Your child's")) board",
                    subtitle: "Big tiles, tap to talk. Best on a tablet in Guided Access.",
                    role: .childBoard
                )
                roleButton(
                    icon: "person.fill",
                    title: "Parent app",
                    subtitle: "Add tiles, start games, see progress, message the board.",
                    role: .parent
                )
            }
            .padding(.horizontal, 20)
            .frame(maxWidth: 560)

            Spacer()
            if opening != nil {
                HStack(spacing: 8) {
                    ProgressView()
                    Text(opening == .childBoard ? "Opening the board — loading pictures & voices…"
                                                : "Opening the parent app…")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(Color(hex: "#ad1457"))
                }
                .padding(.bottom, 24)
            } else {
                Text("You can change this any time in Settings.")
                    .font(.footnote)
                    .foregroundStyle(Color(hex: "#9ca3af"))
                    .padding(.bottom, 24)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color(hex: "#fff7fb"))
    }

    private func roleButton(icon: String, title: String, subtitle: String, role: DeviceMode.Role) -> some View {
        Button {
            guard opening == nil else { return }
            opening = role
            // Let SwiftUI paint the spinner frame BEFORE the heavy destination
            // swap starts, so the tap visibly did something immediately.
            Task { @MainActor in
                await Task.yield()
                mode.role = role
            }
        } label: {
            HStack(spacing: 14) {
                Image(systemName: icon)
                    .font(.system(size: 24, weight: .semibold))
                    .foregroundStyle(.white)
                    .frame(width: 54, height: 54)
                    .background(Color(hex: "#ff1493"), in: Circle())
                VStack(alignment: .leading, spacing: 3) {
                    Text(title)
                        .font(.system(size: 19, weight: .bold, design: .rounded))
                        .foregroundStyle(Color(hex: "#1f2937"))
                        .multilineTextAlignment(.leading)
                        .fixedSize(horizontal: false, vertical: true)
                    Text(subtitle)
                        .font(.system(size: 13))
                        .foregroundStyle(Color(hex: "#6b7280"))
                        .multilineTextAlignment(.leading)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer(minLength: 8)
                if opening == role {
                    ProgressView()
                } else {
                    Image(systemName: "chevron.right")
                        .font(.system(size: 15, weight: .semibold))
                        .foregroundStyle(Color(hex: "#c9b3bf"))
                }
            }
            .padding(18)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(.white, in: RoundedRectangle(cornerRadius: 20))
            .overlay(RoundedRectangle(cornerRadius: 20).stroke(Color(hex: "#f3c6da"), lineWidth: 1))
        }
        .buttonStyle(.plain)
    }
}

File: android-native/app/src/main/java/io/andrewpeterson/myworld/ui/RolePickerView.kt
package io.andrewpeterson.myworld.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.andrewpeterson.myworld.LocalAppContainer
import io.andrewpeterson.myworld.model.DeviceMode
import io.andrewpeterson.myworld.ui.theme.Brand
import io.andrewpeterson.myworld.ui.theme.hexColor

/**
 * Port of `Views/RolePickerView.swift` — first-run "Who uses this device?".
 * Both choices are reversible from settings, so no confirmation step.
 */
@Composable
fun RolePickerView() {
    val c = LocalAppContainer.current
    var opening by remember { mutableStateOf<DeviceMode.Role?>(null) }

    Column(
        Modifier.fillMaxSize().background(hexColor("#fff7fb")).padding(20.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Text("🌍", fontSize = 64.sp)
        Spacer(Modifier.height(12.dp))
        Text("Who uses this device?", fontSize = 30.sp, fontWeight = FontWeight.Bold,
            color = Brand.pinkDeep)
        Spacer(Modifier.height(24.dp))

        val slug = c.auth.user.value?.slug
        val childName = slug?.replace(Regex("peterson$", RegexOption.IGNORE_CASE), "")
            ?.replaceFirstChar { it.uppercaseChar() }?.takeIf { it.isNotBlank() }
        RoleCard(
            emoji = "👆",
            title = "${childName?.plus("'s") ?: "Your child's"} board",
            subtitle = "Big tiles, tap to talk. Best on a tablet with screen pinning.",
            busy = opening == DeviceMode.Role.CHILD_BOARD,
        ) {
            if (opening == null) { opening = DeviceMode.Role.CHILD_BOARD; c.deviceMode.set(DeviceMode.Role.CHILD_BOARD) }
        }
        Spacer(Modifier.height(16.dp))
        RoleCard(
            emoji = "🧑",
            title = "Parent app",
            subtitle = "Add tiles, start games, see progress, message the board.",
            busy = opening == DeviceMode.Role.PARENT,
        ) {
            if (opening == null) { opening = DeviceMode.Role.PARENT; c.deviceMode.set(DeviceMode.Role.PARENT) }
        }

        Spacer(Modifier.height(24.dp))
        if (opening != null) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                CircularProgressIndicator(Modifier.size(18.dp), color = Brand.pinkDeep, strokeWidth = 2.dp)
                Spacer(Modifier.widthIn(min = 8.dp))
                Text(
                    if (opening == DeviceMode.Role.CHILD_BOARD)
                        "  Opening the board, loading pictures & voices…"
                    else "  Opening the parent app…",
                    fontSize = 14.sp, fontWeight = FontWeight.SemiBold, color = Brand.pinkDeep,
                )
            }
        } else {
            Text("You can change this any time in Settings.", fontSize = 13.sp, color = Brand.faint)
        }
    }
}

@Composable
private fun RoleCard(emoji: String, title: String, subtitle: String, busy: Boolean, onTap: () -> Unit) {
    Row(
        Modifier
            .widthIn(max = 560.dp)
            .fillMaxWidth()
            .background(Color.White, RoundedCornerShape(20.dp))
            .border(1.dp, hexColor("#f3c6da"), RoundedCornerShape(20.dp))
            .clickable(onClick = onTap)
            .padding(18.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            Modifier.size(54.dp).background(Brand.pink, CircleShape),
            contentAlignment = Alignment.Center,
        ) { Text(emoji, fontSize = 24.sp) }
        Spacer(Modifier.widthIn(min = 14.dp))
        Column(Modifier.weight(1f).padding(start = 14.dp)) {
            Text(title, fontSize = 19.sp, fontWeight = FontWeight.Bold, color = Brand.ink)
            Text(subtitle, fontSize = 13.sp, color = Brand.muted)
        }
        if (busy) CircularProgressIndicator(Modifier.size(20.dp), color = Brand.pink, strokeWidth = 2.dp)
        else Text("›", fontSize = 22.sp, color = hexColor("#c9b3bf"))
    }
}
